lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,242 @@
+#include "swift/SILOptimizer/Analysis/PostOrderAnalysis.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/Devirtualize.h"
+#include "llvm/ADT/PostOrderIterator.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Debug.h"
+
+using namespace swift;
+
+template<typename...T, typename...U>
+static void diagnose(ASTContext &Context, SourceLoc loc, Diag<T...> diag,
+                     U &&...args) {
+  Context.Diags.diagnose(loc,
+                         diag, std::forward<U>(args)...);
+}
+
+namespace {
+  class DiagnoseInfiniteRecursion : public SILFunctionTransform {
+  public:
+    DiagnoseInfiniteRecursion() {}
+
+  private:
+    static bool hasRecursiveCallInPath(SILBasicBlock &Block,
+                                       SILFunction *Target) {
+      // Process all instructions in the block to find applies that reference
+      // the parent function.  Also looks through vtables for statically
+      // dispatched (witness) methods.
+      auto *TargetModule = Target->getModule().getSwiftModule();
+      for (auto &I : Block) {
+        auto *AI = dyn_cast<ApplyInst>(&I);
+        if (AI && AI->getCalleeFunction() && AI->getCalleeFunction() == Target)
+          return true;
+
+        if (FullApplySite FAI = FullApplySite::isa(&I)) {
+          // Don't touch dynamic dispatch.
+          if (isa<ObjCMethodInst>(FAI.getCallee()))
+            continue;
+
+          auto &M = FAI.getModule();
+          if (auto *CMI = dyn_cast<ClassMethodInst>(FAI.getCallee())) {
+            auto ClassType = CMI->getOperand()->getType();
+
+            // FIXME: If we're not inside the module context of the method,
+            // we may have to deserialize vtables.  If the serialized tables
+            // are damaged, the pass will crash.
+            //
+            // Though, this has the added bonus of not looking into vtables
+            // outside the current module.  Because we're not doing IPA, let
+            // alone cross-module IPA, this is all well and good.
+            auto *BGC = ClassType.getNominalOrBoundGenericNominal();
+            if (BGC && BGC->getModuleContext() != TargetModule) {
+              continue;
+            }
+
+            auto *F = getTargetClassMethod(M, ClassType, CMI);
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+
+          if (auto *WMI = dyn_cast<WitnessMethodInst>(FAI.getCallee())) {
+            SILFunction *F;
+            SILWitnessTable *WT;
+
+            std::tie(F, WT) = M.lookUpFunctionInWitnessTable(
+                WMI->getConformance(), WMI->getMember());
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+        }
+      }
+      return false;
+    }
+
+    using ExitBlockSet = llvm::SmallPtrSetImpl<SILBasicBlock *>;
+    static bool hasInfinitelyRecursiveApply(SILFunction &Fn,
```

Comment:
I don't see any reason for this to be a class member vs. a free-standing static function.